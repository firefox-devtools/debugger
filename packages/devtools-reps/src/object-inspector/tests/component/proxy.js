/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

 /* global jest */

const { mount } = require("enzyme");
const React = require("react");
const { createFactory } = React;
const ObjectInspector = createFactory(require("../../index"));
const { MODE } = require("../../../reps/constants");
const stub = require("../../../reps/stubs/grip").get("testProxy");
const { formatObjectInspector } = require("../test-utils");

const ObjectClient = require("../__mocks__/object-client");
function generateDefaults(overrides) {
  return {
    roots: [{
      path: "root",
      contents: {value: stub}
    }],
    autoExpandDepth: 1,
    mode: MODE.LONG,
    createObjectClient: grip => ObjectClient(grip),
    // Have the prototype already loaded so the component does not call
    // enumProperties for the root's properties.
    loadedProperties: new Map([
      ["root", {prototype: {}}]
    ]),
    ...overrides,
  };
}

function getEnumPropertiesMock() {
  return jest.fn(() => ({
    iterator: {
      slice: () => ({})
    }
  }));
}

describe("ObjectInspector - Proxy", () => {
  it("renders Proxy as expected", () => {
    const enumProperties = getEnumPropertiesMock();

    const props = generateDefaults({
      createObjectClient: grip => ObjectClient(grip, {enumProperties}),
    });
    const oi = mount(ObjectInspector(props));
    expect(formatObjectInspector(oi)).toMatchSnapshot();

    // enumProperties should not have been called.
    expect(enumProperties.mock.calls.length).toBe(0);
  });

  it("calls enumProperties when <target> and <handler> nodes are clicked", () => {
    const enumProperties = getEnumPropertiesMock();

    const props = generateDefaults({
      createObjectClient: grip => ObjectClient(grip, {enumProperties}),
    });
    const oi = mount(ObjectInspector(props));

    const nodes = oi.find(".node");

    const targetNode = nodes.at(1);
    const handlerNode = nodes.at(2);

    targetNode.simulate("click");
    // The function is called twice,  to get  both non-indexed and indexed properties.
    expect(enumProperties.mock.calls.length).toBe(2);
    expect(enumProperties.mock.calls[0][0]).toEqual({ignoreNonIndexedProperties: true});
    expect(enumProperties.mock.calls[1][0]).toEqual({ignoreIndexedProperties: true});

    handlerNode.simulate("click");
    // The function is called twice,  to get  both non-indexed and indexed properties.
    expect(enumProperties.mock.calls.length).toBe(4);
    expect(enumProperties.mock.calls[2][0]).toEqual({ignoreNonIndexedProperties: true});
    expect(enumProperties.mock.calls[3][0]).toEqual({ignoreIndexedProperties: true});
  });
});
