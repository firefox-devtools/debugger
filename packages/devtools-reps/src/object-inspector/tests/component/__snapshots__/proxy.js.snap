// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectInspector - Proxy renders Proxy as expected 1`] = `
"
▼ Proxy
|  ▶︎ <target>: Object { … }
|  ▶︎ <handler>: Array(3) [ … ]
"
`;
